File: OOPRecords.Model/SubjectRepository.cs
using NakedObjects;
using System.Linq;

namespace OOPRecords.Model
{
    public class SubjectRepository
    {
        public IDomainObjectContainer Container { set; protected get; }

        public Subject CreateNewSubject()
        {
            return Container.NewTransientInstance<Subject>();
        }

        public IQueryable<Subject> AllSubjects()
        {
            return Container.Instances<Subject>();
        }

        public IQueryable<Subject> FindSubjectByName(string name)
        {
            return AllSubjects().Where(c => c.Name.ToUpper().Contains(name.ToUpper()));
        }
    }
}

File: OOPRecords.ModelVB/SubjectRepository.vb
Imports NakedObjects

Public Class SubjectRepository

    Private _container As IDomainObjectContainer
    Public WriteOnly Property Container() As IDomainObjectContainer
        Set(ByVal value As IDomainObjectContainer)
            _container = value
        End Set
    End Property

    Public Function CreateNewSubject() As Subject
        Return _container.NewTransientInstance(Of Subject)()
    End Function

    Public Function AllSubjects() As IQueryable(Of Subject)
        Return _container.Instances(Of Subject)()
    End Function

    Public Function FindSubjectByName(ByVal name As String) As IQueryable(Of Subject)
        Return AllSubjects().Where(Function(c) c.Name.ToUpper().Contains(name.ToUpper()))
    End Function
End Class
